File: scripts/graph_rag/tsparser/src/integrations.ts
import { SourceFile } from "ts-morph";

export function detectIntegrations(src: SourceFile) {
  const text = src.getFullText();
  const mongo = new Set<string>();
  const natsPub = new Set<string>();
  const natsSub = new Set<string>();
  const env = new Set<string>();

  // Mongoose: mongoose.model("Name", schema) or model("Name", schema)
  for (const m of text.matchAll(/(?:mongoose\.)?model\(\s*["']([A-Za-z][A-Za-z0-9_]*)["']/g)) {
    mongo.add(m[1]);
  }
  // db.collection("name") / getCollection("name")
  for (const m of text.matchAll(/(?:getCollection|\.collection)\(\s*["']([A-Za-z][A-Za-z0-9_]*)["']/g)) {
    mongo.add(m[1]);
  }
  // @Schema({ collection: "name" })
  for (const m of text.matchAll(/collection\s*:\s*["']([A-Za-z][A-Za-z0-9_]*)["']/g)) {
    mongo.add(m[1]);
  }

  // NATS: jsCtx.publish("subject", ...) / nc.publish("subject", ...)
  for (const m of text.matchAll(/\.publish\(\s*["']([a-zA-Z0-9_.\-\*]+)["']/g)) {
    natsPub.add(m[1]);
  }
  // consume({ subject: "..." }) / subscribe("...")
  for (const m of text.matchAll(/\.subscribe\(\s*["']([a-zA-Z0-9_.\-\*]+)["']/g)) {
    natsSub.add(m[1]);
  }
  for (const m of text.matchAll(/subject\s*:\s*["']([a-zA-Z0-9_.\-\*]+)["']/g)) {
    natsSub.add(m[1]);
  }

  // process.env.FOO / process.env["FOO"]
  for (const m of text.matchAll(/process\.env\.([A-Z][A-Z0-9_]*)/g)) env.add(m[1]);
  for (const m of text.matchAll(/process\.env\[["']([A-Z][A-Z0-9_]*)["']\]/g)) env.add(m[1]);

  return {
    mongo: [...mongo],
    nats_pub: [...natsPub],
    nats_sub: [...natsSub],
    env: [...env],
  };
}

File: scripts/graph_rag/pyparser/pyparser/integrations.py
"""Regex-driven extraction of Mongo collections, NATS subjects, env vars."""
from __future__ import annotations

import re


_MONGO_COLL = re.compile(
    r"""(?:                       # matchers for collection name literals
      db\[[\"']([A-Za-z_][A-Za-z0-9_]*)[\"']\]
    | \.get_collection\([\"']([A-Za-z_][A-Za-z0-9_]*)[\"']\)
    | database\[[\"']([A-Za-z_][A-Za-z0-9_]*)[\"']\]
    | \.collection\([\"']([A-Za-z_][A-Za-z0-9_]*)[\"']\)
    )""",
    re.VERBOSE,
)

_NATS_PUB = re.compile(r"""\.publish\(\s*[\"']([a-zA-Z0-9_.\-\*]+)[\"']""")
_NATS_SUB = re.compile(r"""\.subscribe\(\s*[\"']([a-zA-Z0-9_.\-\*]+)[\"']""")

_ENV = re.compile(r"""os\.(?:environ\.get|environ|getenv)\(\s*[\"']([A-Z][A-Z0-9_]*)[\"']""")


def detect_integrations(src: str) -> dict:
    mongo = set()
    for m in _MONGO_COLL.finditer(src):
        for g in m.groups():
            if g:
                mongo.add(g)
    nats_pub = {m.group(1) for m in _NATS_PUB.finditer(src)}
    nats_sub = {m.group(1) for m in _NATS_SUB.finditer(src)}
    env = {m.group(1) for m in _ENV.finditer(src)}
    return {
        "mongo": sorted(mongo),
        "nats_pub": sorted(nats_pub),
        "nats_sub": sorted(nats_sub),
        "env": sorted(env),
    }
